src/hosting/proxy.ts

```diff
@@ -35,67 +41,94 @@ function makeVary(vary?: string): string {
  * the Firebase Hosting origin.
  */
 export function proxyRequestHandler(url: string, rewriteIdentifier: string): RequestHandler {
-  return (req: Request, res: Response, next: () => void): any => {
+  return async (req: IncomingMessage, res: ServerResponse, next: () => void): Promise<void> => {
     logger.info(`[hosting] Rewriting ${req.url} to ${url} for ${rewriteIdentifier}`);
     // Extract the __session cookie from headers to forward it to the
     // functions cookie is not a string[].
     const cookie = (req.headers.cookie as string) || "";
     const sessionCookie = cookie.split(/; ?/).find((c: string) => {
-      return c.trim().indexOf("__session=") === 0;
+      return c.trim().startsWith("__session=");
     });
 
-    const proxied = request({
-      method: req.method,
-      qs: req.query,
-      url: url + req.url,
-      headers: {
-        "X-Forwarded-Host": req.headers.host,
-        "X-Original-Url": req.url,
-        Pragma: "no-cache",
-        "Cache-Control": "no-cache, no-store",
-        // forward the parsed __session cookie if any
-        Cookie: sessionCookie,
-      },
-      followRedirect: false,
-      timeout: 60000,
-    });
+    // req.url is just the full path (e.g. /foo?key=value; no origin).
+    const u = new URL(url + req.url);
+    const c = new Client({ urlPrefix: u.origin, auth: false });
+    const controller = new AbortController();
+    const timer: NodeJS.Timeout = setTimeout(() => controller.abort(), 60000);
```

Comment:
nit: Consider using `Promise.race` with a Promise that resolves in 60s to make it less scary w.r.t race conditions

Reply:
Ah nevermind, I see how abort errors are handled before pipes now